src/Compilers/Core/Portable/Text/ChangedText.cs

```diff
@@ -27,8 +25,50 @@ public ChangedText(SourceText oldText, SourceText newText, ImmutableArray<TextCh
             Debug.Assert(!changeRanges.IsDefault);
 
             _newText = newText;
-            _weakOldText = new WeakReference<SourceText>(oldText);
-            _changes = changeRanges;
+            _info = new ChangeInfo(changeRanges, new WeakReference<SourceText>(oldText), (oldText as ChangedText)?._info);
+        }
+
+        private class ChangeInfo
+        {
+            public ImmutableArray<TextChangeRange> ChangeRanges { get; }
+
+            // store old text weakly so we don't form unwanted chains of old texts (especially chains of ChangedTexts)
+            // used to identify the changes in GetChangeRanges.
+            public WeakReference<SourceText> WeakOldText { get; }
+
+            public ChangeInfo Previous { get; private set; }
```

Comment:
OldText gets to get reclaimed by GC, so we cannot use it to track backward.   Imagine a scenario with doc versions a, b, c, and d.   You have references to original 'a' and final 'd', but 'b' and 'c' are forgotten for now.  The GC may collect them, and that is good and intended, but as long 'a' is still alive you'd like to be able to compute the differences between 'a' and 'd'.  This is the motivating purpose for this PR.